src/com/google/cose/Encrypt0Message.java

```diff
@@ -99,4 +121,26 @@ public byte[] getCiphertext() {
   public static Builder builder() {
     return new Builder();
   }
+
+  public byte[] decrypt(EncryptionKey key, byte[] detachedCiphertextContent, byte[] externalAad,
+      Algorithm algorithm) throws CborException, CoseException {
+    byte[] ciphertext = this.ciphertext;
+    if (ciphertext == null) {
+      ciphertext = detachedCiphertextContent;
+    }
+
+    // find algorithm in the unprotected headers if provided with null.
+    if (algorithm == null) {
+      algorithm = Algorithm.fromCoseAlgorithmId(
+          CborUtils.asInteger(findAttributeInHeaders(Headers.MESSAGE_HEADER_ALGORITHM))
```

Comment:
RFC 8152 says:
```
alg:  This parameter is used to indicate the algorithm used for the
      security processing.  This parameter MUST be authenticated where
      the ability to do so exists.  This support is provided by AEAD
      algorithms or construction (COSE_Sign, COSE_Sign0, COSE_Mac, and
      COSE_Mac0).  This authentication can be done either by placing the
      header in the protected header bucket or as part of the externally
      supplied data.
```

So I don't think this is valid. Either we read `alg` from the protected headers or not at all.

Reply:
but wouldn't provided algorithm be considered part of the externally supplied data?
or do you mean we should not be searching this in unprotected data?

Reply:
My interpretation is that `alg` must always be authenticated, so there are two ways to do this:
1. It's in a protected header
2. It's in externally-supplied data (probably included as part of AAD)

It should never be in unprotected headers.

Reply:
My understanding was that if the algorithm is an AEAD algorithm, then it can be passed in as an algorithm directly.
I agree with the unprotected header portion and will remove the search in the unprotected headers.